[mono] Reorganize zlib dependency logic

lambdageek opened the pull request:
Instead of deciding whether features are enabled based on the presence or absence of zlib, add switches to control features (embedded PDB support, and compressed log profiler output) and based on that require zlib.

On some platforms we use the in-tree copy of zlib from src/native/external/zlib, otherwise we want to link against the system zlib.

Previously, if zlib was missing the build would quietly succeed, but the feature support would be missing.  Now, if the feature isn't disabled but zlib is missing, we fail the build.

If both features are disabled, don't depend on zlib at all.

In particular this re-enables embedded PDB support on iOS and Android

The corresponding release/6.0-maui PR is #63359 and is more conservative - it just duplicates the mono/mono HAVE_SYS_ZLIB check.   In this PR, we get rid of `HAVE_SYS_ZLIB` entirely.

lambdageek committed:
[mono] Add back HAVE_SYS_ZLIB support

In particular this re-enables embedded PDB support on iOS and Android

lambdageek commented on src/mono/cmake/configure.cmake:
```diff
@@ -94,6 +94,21 @@ endif()
+  #include <zlib.h>
+  int main(void)
+  {
+    #if defined(ZLIB_VERNUM) && (ZLIB_VERNUM >= 0x1230)
+    #else
+    #error No good zlib found
+    #endif
+    return 0;
```
This is copied from

https://github.com/mono/mono/blob/a791978a6e4312698137be451c648b650a8c8663/configure.ac#L728-L732

but I'm not sure if that's necessary.

am11 replied:
We can include (`include(${CLR_SRC_NATIVE_DIR}/libs/System.IO.Compression.Native/extra_libs.cmake)`), call (`set(Z_LIBS)\nappend_extra_compression_libs(Z_LIBS)`) this macro: https://github.com/dotnet/runtime/blob/60012a1092e6cb0f4fe6fc9d9ffd9907826a8355/src/native/libs/System.IO.Compression.Native/extra_libs.cmake#L2  and skip the introspection check. The difference is that in libs, _when_ we require zlib package on the build machine, we fail the build if it was not found. In mono, we treat it as optional dependency. It would be nice to streamline the requirements and cleanup `!build_from_source && !zlib_found` fallback code.

lambdageek replied:
@am11 Yea I think this probably needs a bit more work for net7.  I think the logic in Mono is kind of backwards: we probe for zlib (or in some cases statically link the in-tree copy) and based on that we enable or disable certain features in the runtime (Embedded PDB support; compressed log profiler output).  I think that's backwards.  We should have feature options and those should drive the zlib dependency decisions.

I'll rework it and include the cmake fragment from libs.

am11 replied:
A note on another difference between mono and rest of the native partitions in repo: all native partitions use `eng/native/cofigureplatform.cmake`, which defines `CLR_CMAKE_{HOST, TARGET}_*` properties. In mono, we only use configure*.cmake in `src/mono/dlls/*`, for the remaining areas, we define a separate set of platform properties. In future, if we change mono to also use common properties, it will be possible to share more configuration code without reinterpreting the semantics.

As it stands, `append_extra_compression_libs` has these differences:

|  rest of the repo | mono |
| --- | --- |
| `CLR_CMAKE_TARGET_BROWSER` | `HOST_BROWSER` |
| `CLR_CMAKE_TARGET_ANDROID` | `HOST_ANDROID` |
| `CLR_CMAKE_TARGET_SUNOS` | `HOST_SOLARIS` |

lambdageek replied:
Oh that's annoying. 

Do you think it makes sense to modify `append_extra_compression_libs` to also check for the mono names, @am11?

(In the long term we should use the `CLR_CMAKE_` names everywhere, and change mono's `HOST_ABC` macros to `CLR_CMAKE_TARGET_ABC` and mono's `TARGET_XYZ` macros to something like `CODEGEN_TARGET_XYZ` - mono uses target to mean the platform for which the JIT or AOT compiler is generating code)

am11 replied:
> Do you think it makes sense to modify `append_extra_compression_libs` to also check for the mono names, @am11?

Yup, that will work. `if (CLR_CMAKE_TARGET_BROWSER OR HOST_BROWSER)` etc.

> mono uses target to mean the platform for which the JIT or AOT compiler is generating code

I understand this difference and hope we don't carry it in the future by using same semantics everywhere. :slightly_smiling_face:
On a closer look, HOST and TARGET have different meaning in mono vs. other repo partitions. Earlier, Steve streamlined the difference in for other partitions in https://github.com/dotnet/runtime/pull/1974. It is mostly how VS / Windows use the terms. Mono was originally using Unix-y autoconf, which has a bit different meanings; `Host` is what VS calls `Target` and `Build` is what VS refers to as `Host`. `Target` in autoconf (and hence mono) is a special case, where compiler itself is built on a different architecture than `Host` and `Build`, and `TARGET` is used to indicate the machine for which code is generated:
  
* https://www.gnu.org/software/autoconf/manual/autoconf-2.65/html_node/Specifying-Target-Triplets.html
* https://stackoverflow.com/a/15901574

lambdageek replied:
> I understand this difference and hope we don't carry it in the future by using same semantics everywhere. 

yea I agree.  (it's my long (loooong) term hope that mono's AOT compiler can be made multitargeting so the autoconf-sense `target` will go away entirely as a build-time idea and we can pick the AOT output triple when we invoke the AOT compiler)

lambdageek commented:
/azp run runtime-manual

lambdageek commented:
See https://github.com/dotnet/runtime/pull/63365#discussion_r778854216

I think we can do better for net7: have explicit options for enable/disable Embedded PDB support and compressed log profiler output, and use those to decide whether zlib is required or not, and if it is required, whether we should depend on the system zlib, or the in-tree copy.

lambdageek committed:
[mono] Reorganize zlib dependency logic

Instead of deciding whether features are enabled based on the presence or
absense of zlib, add switches to control features (embedded PDB support, and
compressed log profiler output) and based on that require zlib.

On some platforms we use the in-tree copy of zlib from
src/native/external/zlib, otherwise we want to link against the system zlib.

Previously, if zlib was missing the build would quietly succeed, but the
feature support would be missing.  Now, if the feature isn't disabled but zlib
is missing, we fail the build.

If both features are disabled, don't depend on zlib at all.

lambdageek committed:
Fix typos

Co-authored-by: Adeel Mujahid <3840695+am11@users.noreply.github.com>

lambdageek committed:
Style nits

- reverse if() logic to avoid negation

- use cmakedefine without a value in config.h.in, same as CoreCLR
